subprojects/core/src/main/java/org/gradle/caching/internal/controller/impl/RootBuildCacheControllerRef.java

```diff
@@ -17,63 +17,179 @@
 package org.gradle.caching.internal.controller.impl;
 
 import org.gradle.caching.BuildCacheKey;
+import org.gradle.caching.configuration.internal.BuildCacheConfigurationInternal;
 import org.gradle.caching.internal.CacheableEntity;
 import org.gradle.caching.internal.controller.BuildCacheController;
+import org.gradle.caching.internal.controller.NoOpBuildCacheController;
 import org.gradle.caching.internal.controller.service.BuildCacheLoadResult;
+import org.gradle.caching.internal.services.BuildCacheControllerFactory;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.instantiation.InstanceGenerator;
+import org.gradle.internal.service.scopes.Scope;
+import org.gradle.internal.service.scopes.ServiceScope;
 import org.gradle.internal.snapshot.FileSystemSnapshot;
+import org.gradle.util.Path;
 
+import java.io.IOException;
 import java.time.Duration;
 import java.util.Map;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
 
+/**
+ * This service manages the state of the {@link BuildCacheController} instances in the build tree.
+ * A separate instance is created for each build in the tree. This instance delegates to another immutable controller based on the state of the root build.
+ *
+ * <p>The implementation provides the following behavior:
+ *
+ * <ul>
+ *     <li>All builds in the tree use the root build's cache configuration, once that configuration is available (ie the root build's settings have been evaluated).</li>
+ *     <li>If caching is required in non-root build before the root build's configuration is available, that build's own cache configuration will be used.
```

Comment:
Assuming an early included build with remote cache configured differently than the root build, it's unclear which units of work will be loaded from the included build's remote cache, and which from that of the root build. I guess this is an acceptable compromise for now, but it'd be good to explicitly document the limitation here.

Reply:
Added some Javadoc.

For this PR, we're not trying to solve the problem of knowing when in the lifecycle a unit of work will run. This was already a problem before the changes.